### ISSUE TITLE
Allow returning Vec<T>

### ISSUE THREAD

**Issue by rookboom** [rustwasm/wasm-bindgen#111]
It would really be great to be able to return a vector of structs or tuples:

For example. I have the following type:
```
#[wasm_bindgen]
struct Range {
    offset: u32,
    length: u32
}
```

that I want to return in my function

```
#[wasm_bindgen]
pub fn token_ranges(text: &str) -> Vec<Range> 
```

I am getting this error:

```
   Compiling picl_wasm_runtime v0.1.0 (file:///A:/Repos/picl_native_runtime/bindings/typescript)
error[E0277]: the trait bound `std::boxed::Box<[Range]>: wasm_bindgen::convert::WasmBoundary` is not satisfied
  --> src\lib.rs:15:1
   |
15 | #[wasm_bindgen]
   | ^^^^^^^^^^^^^^^ the trait `wasm_bindgen::convert::WasmBoundary` is not implemented for `std::boxed::Box<[Range]>`
   |
   = help: the following implementations were found:
             <std::boxed::Box<[u16]> as wasm_bindgen::convert::WasmBoundary>
             <std::boxed::Box<[i16]> as wasm_bindgen::convert::WasmBoundary>
             <std::boxed::Box<[f32]> as wasm_bindgen::convert::WasmBoundary>
             <std::boxed::Box<[i32]> as wasm_bindgen::convert::WasmBoundary>
           and 5 others
   = note: required because of the requirements on the impl of `wasm_bindgen::convert::WasmBoundary` for `std::vec::Vec<Range>`
```

My workaround is to flatten my data and return a Vec<u32> and then splice my token ranges on the JS side. This is unfortunate...

How can I add a WasmBoundary trait for a custom type?
Is there a better way?

**Comment by alexcrichton** [comment]
Thanks for the report!

Right now there's not a great way to do this in wasm-bindgen itself in terms of the conversion here has to be also accompanied with a conversion on the JS side which may be somewhat tricky. It should certainly be possible though given enough support!

**Comment by Hywan** [comment]
Same issue here :-).

What kind of support do you need?

**Comment by alexcrichton** [comment]
@rookboom @Hywan do y'all basically need `Vec<T>` where `T` has `#[wasm_bindgen]` on it?

**Comment by Hywan** [comment]
@alexcrichton Exactly, that's precisely my usecase.

**Comment by rookboom** [comment]
Yes, that would be fantastic.

**Comment by alexcrichton** [comment]
Ok this is then I think pretty similar to https://github.com/rustwasm/wasm-bindgen/issues/104. We somehow need a way to communicate this to the CLI tool but currently the strategy for doing that is quite limited.

**Comment by gabrielcarneiro97** [comment]
I'm with a similar issue here, but with a complex struct.

code:
```
#[wasm_bindgen]
pub struct Cell {
    state: State,
    position: Coords,
    neighboors: Vec<Coords>,
    neighboors_alive: i32
}

#[wasm_bindgen]
impl Cell {
    pub fn new(state: State, position: Coords, neighboors: Vec<Coords>) -> Cell {
        Cell {
            state,
            position,
            neighboors,
            neighboors_alive: 0
        }
    }
}
```

error:
```
error[E0277]: the trait bound `std::boxed::Box<[Coords]>: wasm_bindgen::convert::FromWasmAbi` is not satisfied
  --> src\lib.rs:36:1
   |
36 | #[wasm_bindgen]
   | ^^^^^^^^^^^^^^^ the trait `wasm_bindgen::convert::FromWasmAbi` is not implemented for `std::boxed::Box<[Coords]>`
   |
   = help: the following implementations were found:
             <std::boxed::Box<[u16]> as wasm_bindgen::convert::FromWasmAbi>
             <std::boxed::Box<[wasm_bindgen::JsValue]> as wasm_bindgen::convert::FromWasmAbi>
             <std::boxed::Box<[u8]> as wasm_bindgen::convert::FromWasmAbi>
             <std::boxed::Box<[f32]> as wasm_bindgen::convert::FromWasmAbi>
           and 5 others
   = note: required because of the requirements on the impl of `wasm_bindgen::convert::FromWasmAbi` for `std::vec::Vec<Coords>`
```

**Comment by David-OConnor** [comment]
Gabriel and rook - have y'all found a workaround? Solving or working around this would add much flexibility to WASM in Rust.

**Comment by rookboom** [comment]
My  workaround is to pass JSON over the wasm boundary... Not ideal but works for now.

**Comment by Gisleburt** [comment]
For those looking for a workaround on this, if you can turn your data into a `Vec<u8>` or `&[u8]`

```rust
#[wasm_bindgen]
pub struct ByteStream {
    offset: *const u8,
    size: usize,
}

#[wasm_bindgen]
impl ByteStream {
    pub fn new(bytes: &[u8]) -> ByteStream {
        ByteStream {
            offset: bytes.as_ptr(),
            size: bytes.len(),
        }
    }

    pub fn offset(&self) -> *const u8 {
        self.offset
    }

    pub fn size(&self) -> usize {
        self.size
    }
}
```
A good example of how to use this is creating a texture in Rust to render in Javascript, so for example: 
```rust
pub fn render() -> ByteStream {
  let texture = Vec::new();
  // ...
  ByteStream::new(&texture);
}

```

```js
const texture = render();
const textureRaw = new Uint8ClampedArray(memory.buffer, texture.offset(), texture.size());
const image = new ImageData(textureRaw, width, height);
```

**Comment by fitzgen** [comment]
Rather than returning pointers and lengths manually, you can use this, which should be slightly less error prone: https://docs.rs/js-sys/0.3.9/js_sys/struct.Uint8Array.html#method.view

**Comment by lyd600lty** [comment]
@Gisleburt I try for this, but build `ByteStream` with offset and length will broke the memory in my test.
![image](https://user-images.githubusercontent.com/2917899/54975969-cbf47480-4fd3-11e9-9a64-d7316b595c25.png)

**Comment by stefan2718** [comment]
If anyone is still looking at this, I was able to work around this using Serde to serialize/deserialize the data. This was the guide I used: https://rustwasm.github.io/docs/wasm-bindgen/reference/arbitrary-data-with-serde.html

**Comment by theashguy** [comment]
I'm wanting to take this one step further and return Vec<js_sys::Function> which means no serde serialization for me. The workaround I'm kicking about at the moment is to create the collection on the JS side and expose some methods for appending and cleaning up.

Something like...

```
window.vecCache = {}

window.vecCacheAppend = function (key, item) {
  if(!window.vecCache[key]) {
    window.vecCache[key] = []
  }

  window.vecCache[key].push(item)
}

window.vecCacheClear = function (key) {
  delete window.vecCache[key]
}
```

**Not suitable for prod work** and by no means ideal or ergonomic but I'm really just thrashing around to see how far I can take `rustwasm` at the moment :D. Possibly an idea for anyone blocked by this issue and wanting to do things more complex. Also an extra AC for the team to add to the backlog!

**Comment by Zireael07** [comment]
This issue is kind of a blocker for pretty much any sort of a bigger project using rustwasm.

**Comment by Pauan** [comment]
It's not a complete solution, but I created #1749 which adds in `FromIterator` for `Array`:

```rust
use js_sys::Array;

#[wasm_bindgen]
pub fn token_ranges(text: &str) -> Array {
    get_vec_somehow().into_iter().collect()
}
```

This means that now you *can* send `Vec<T>` to JS, you just have to return an `Array` and use `.into_iter().collect()` to convert the `Vec<T>` into an `Array`.

**Comment by dasoncheng** [comment]
Same issue here :-).
```rs
#[wasm_bindgen]
#[derive(Debug)]
pub struct HeartBeat {
    pub template: u8,
    pub classify: u8,
    pub index: u32,
    pub tr: u16,
    pub hr: u16,
    pub feature: [[f32; 20]; 3],
}
```

```error
the trait bound `[[f32; 20]; 3]: wasm_bindgen::convert::traits::IntoWasmAbi` is not satisfied

the trait `wasm_bindgen::convert::traits::IntoWasmAbi` is not implemented for `[[f32; 20]; 3]`

help: the following implementations were found:
        <&'a [f32] as wasm_bindgen::convert::traits::IntoWasmAbi>
        <&'a [f64] as wasm_bindgen::convert::traits::IntoWasmAbi>
        <&'a [i16] as wasm_bindgen::convert::traits::IntoWasmAbi>
        <&'a [i32] as wasm_bindgen::convert::traits::IntoWasmAbi>
      and 20 othersrustc(E0277)
lib.rs(38, 1): the trait `wasm_bindgen::convert::traits::IntoWasmAbi` is not implemented for `[[f32; 20]; 3]`
```

**Comment by dragly** [comment]
@Pauan Great! Thanks for looking into this.

However, I am encountering the following error now.

```
the trait `std::convert::AsRef<wasm_bindgen::JsValue>` is not implemented for `MyObject`
note: required because of the requirements on the impl of `std::iter::FromIterator<MyObject>` for `js_sys::Array`
```

from the following code:

```rust
#[wasm_bindgen]
struct MyObject {
    a: f32,
    b: f32,
}

#[wasm_bindgen]
pub fn test() -> Array {
    let objects = vec![
        MyObject {
            a: 123.0,
            b: 1024.1,
        },
        MyObject {
            a: 456.7,
            b: 1024.8,
        },
    ];
    objects.into_iter().collect()
}
```
Am I missing something?

This is using wasm-bindgen = "0.2.51". Maybe your change is not in yet? The error is different from before, however, so it seems like something changed.

**Comment by Pauan** [comment]
@dragly As explained in the PR, you need to use `.map(JsValue::from)`, like this:

```rust
objects.into_iter().map(JsValue::from).collect()
```

This is because structs are a Rust data type, and so you have to manually use `JsValue::from` to convert them into a JS data type (the same is true for other Rust data types like `&str`, `i32`, etc.).

**Comment by Kinrany** [comment]
As a workaround, is it possible to also specify the TypeScript type returned by the function?

By default `fn foo() -> Array` is compiled into `() => any[]` instead of, say, `string[]`.

**Comment by Pauan** [comment]
@Kinrany Yes, but it requires a bit of a hack:

```rust
#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(typescript_type = "Array<string>")]
    type MyArray;
}
```

(The `typescript_type` attribute can specify any TypeScript type, even complex types like `|`)

Now you just do `fn foo() -> MyArray` and use `.unchecked_into::<MyArray>()` to cast the `Array` into `MyArray`.

**Comment by tlaukkan** [comment]
It would still be awesome to have builtin support for something like this which is quite normal case:
```
#[wasm_bindgen]
pub struct IntersectResult {
    pub hit: bool,
    pub triangle_index: u32,
    pub u: f32,
    pub v: f32,
    pub distance: f32,
}

#[wasm_bindgen]
pub struct IntersectResultArray {
    pub intersects: Vec<IntersectResult>,
}
```

**Comment by valentinpi** [comment]
Any updates on this? Natively using fixed-size arrays with WASM would make this so much more practicable.

`pub struct SChip8 {
    pc: usize,                           //
    ar: u16,                             // Address register
    sp: usize,                           //
    r: [u8; 8],                          // RPL Flags
    v: [u16; 16],                        //
    pub dt: u8,                          // Delay timer
    pub st: u8,                          // Sound timer
    stack: [usize; 48],                  // Stack implemented as empty ascending
    ram: [u8; 0x1000],                   //
    pub screen: [u8; SCHIP8_NUM_PIXELS], //
    pub screen_width: usize,             //
    pub screen_height: usize,            //
    pub extended_screen: bool,           //
    pub key_pad: [bool; 16],             //
}`

**Comment by i-schuetz** [comment]
What do we have to do currently to achieve this? E.g to pass `Vec<Foo>` to JS:
```rust
struct Foo {
    field: i32,
    str: String,
}
```

**Comment by fjarri** [comment]
@i-schuetz : the workaround that I use is having a return type of `Vec<JsValue>` and converting your vector to it on return: `myvec.iter().map(JsValue::from).collect()`.

**Comment by fjarri** [comment]
Does this issue also cover the support for `Vec<T>` function arguments? Because that's also impossible at the moment, and much harder to work around.

**Comment by ppershing** [comment]
Unfortunately, solutions with `js_sys` have huge performance overheads.
You can also return `Box<[T]>` with `T` being *basic numeric type* which is efficient on the JS side (it just slices relevant wasm memory). This however needs a copy on the Rust side. In general, it is faster than `js_sys` though.

What I come up (and seems to be faster by a margin) is to have the following wrapper type:

```
pub struct MySlice<T> {
    phantom: std::marker::PhantomData::<T>,
    _ptr: u32,
    _len: u32,
}

impl<T: wasm_bindgen::describe::WasmDescribe> wasm_bindgen::describe::WasmDescribe for MySlice<T> {
    fn describe() {
        wasm_bindgen::describe::inform(wasm_bindgen::describe::REF);
        wasm_bindgen::describe::inform(wasm_bindgen::describe::SLICE);
        T::describe();
    }
}

impl<T: wasm_bindgen::describe::WasmDescribe> wasm_bindgen::convert::IntoWasmAbi for MySlice<T> {
    type Abi=wasm_bindgen::convert::WasmSlice;

    #[inline]
    fn into_abi(self) -> wasm_bindgen::convert::WasmSlice {
        wasm_bindgen::convert::WasmSlice {
            ptr: self._ptr,
            len: self._len,
        }
    }
}

impl<T> std::convert::From<&Vec<T>> for MySlice<T> {
    fn from(vec: &Vec<T>) -> Self {
        let _ptr = vec.as_ptr() as u32;
        let _len = vec.len() as u32;
        
        Self {
            phantom: std::marker::PhantomData,
            _ptr,
            _len
        }
    }
}
```

This is both zero-copy on the rust side and single copy between wasm memory and JS on the JS side.
Basically, the type just emulates ref to a slice through standard wasm bindgen API (which means support for returning `&[T]` should be relatively easy to add if somebody knows internals of bindgen)
A caveat is that you *shouldn't* save MySlice, it should only be used to return immediate data to JS (because it detaches pointers, the code cannot guarantee lifetimes)

Usage is simple:
```
#[wasm_bindgen]
pub struct Container {
    data: Vec<f64>,
}

#[wasm_bindgen]
impl Container {
    pub fn read(&self) -> MySlice<f64> {
        (&self.data).into()
    }
}
```

with the JS side using
```
const c = Container.new() // somehow create
const values = c.read()
```

**Comment by Zireael07** [comment]
> You can also return Box<[T]> with T being basic numeric type 

This is about Vec<T> where T is a struct of some sort, so how is that relevant?

**Comment by Elias-Graf** [comment]
> Kinrany Yes, but it requires a bit of a hack:
> 
> ```rust
> #[wasm_bindgen]
> extern "C" {
>     #[wasm_bindgen(typescript_type = "Array<string>")]
>     type MyArray;
> }
> ```
> 
> (The `typescript_type` attribute can specify any TypeScript type, even complex types like `|`)
> 
> Now you just do `fn foo() -> MyArray` and use `.unchecked_into::<MyArray>()` to cast the `Array` into `MyArray`.

@Pauan How would one go about receiving such a `MyArray` type in another function? How do I parse it back to a rust slice so I can work with it?

E.g.:
```rust
pub fn processs_my_array(arr: MyArray) -> SomeNewThing {
    // How to convert back to [string] or rather some more complicated struct slice?
}
```

**Comment by Pauan** [comment]
@Elias-Graf After you convert the `Vec` into an `Array` (or a `MyArray`) it is now a JS value, it's no longer a Rust value.

So you have to process it using the [JS `Array` methods](https://docs.rs/js-sys/0.3.52/js_sys/struct.Array.html):

```rust
pub fn processs_my_array(arr: MyArray) -> SomeNewThing {
    let arr: js_sys::Array = arr.unchecked_into();

    // If you only need an Iterator then you don't need to use collect
    let strings: Vec<String> = arr.iter().map(|x| x.as_string().unwrap()).collect();
}
```

As for converting `JsValue` into structs... that is a separate issue:

https://github.com/rustwasm/wasm-bindgen/issues/1642

https://github.com/rustwasm/wasm-bindgen/issues/2231

**Comment by zimond** [comment]
I believe latest wasm-bindgen release already allows return `Vec<T>` where `T` is any type implementing `WasmDescribe + JsCast`. Currently available types includes primitives, and types from `js-sys`. So returning `Vec<JsValue>` is fine now.

**Comment by PedroFonsecaDEV** [comment]
@zimond 

Do you have a simple example of how to implement this?

Let's suppose that your function returns a Vector of numbers; how to implement the `WasmDescribe + JsCast` in order to make it work?

Thanks.

**Comment by zimond** [comment]
@PedroFonsecaDEV 
`WasmDescribe` is already implemented for simple types (like numbers), code [here](https://github.com/rustwasm/wasm-bindgen/blob/master/src/describe.rs)

For complex types like strings and structs, I still don't know a way to make this work.

**Comment by PedroFonsecaDEV** [comment]
@zimond Yes, sorry my bad. 

@Pauan @alexcrichton 
Do you guys have a simple example which we could use as a guide to implement a function that returns `Vec<Struct>` ?

**Comment by alexcrichton** [comment]
I don't believe that's supported today. That's what this issue is about though.

**Comment by mbround18** [comment]
I was able to get this solution working for me, as a disclaimer, I just hacked it together and it could be greatly improved but I wanted to get something working before i attempted to clean it up into macros. 

`Cargo.toml`
```toml
[dependencies]
serde = { version = "1", features = ["derive"] }
wasm-bindgen = {version= "0", features=["serde-serialize"]}
js-sys = "0.3.55"
```

`lib.rs`
```rust
// Struct Code
#[wasm_bindgen]
#[derive(Serialize, Deserialize)]
struct User {
    id: String,
    name: String,
    age: i64
}

impl From<std::option::Option<query_user::QueryUserQueryUser>> for User {
    fn from(user: Option<query_user::QueryUserQueryUser>) -> Self {
        let data = user.as_ref().expect("user");
        User {
            id: String::from(&data.id),
            name: String::from(data.name.as_ref().expect("name")),
            age: data.age.expect("age")
        }
    }
}

// Excluded the unrelated code needed here
#[wasm_bindgen]
pub async fn get_user() -> js_sys::Array {
  // Code that fires off a request and sets response
  response.data
    .expect_throw("response data")
    .query_user 
    .expect_throw("query_user").into_iter().map(User::from).map( { |user|
        JsValue::from_serde(&user).unwrap()
    }).collect::<js_sys::Array>()
}
```

![image](https://user-images.githubusercontent.com/12646562/141877492-a7d9ed76-b331-465c-bc82-1b270c1a79a0.png)

**Comment by sinking-point** [comment]
The docs seem to incorrectly indicate that this is already supported: https://rustwasm.github.io/wasm-bindgen/reference/types/boxed-jsvalue-slice.html

> Boxed slices of imported JS types and exported Rust types are also supported. Vec<T> is supported wherever Box<[T]> is.

**Comment by shral** [comment]
Similar problem here, I have a HasMap containing strings and Files (<String, Option<web_sys::File>>) want to return that to JavaScript or at least some kind of array of the Files

**Comment by 9oelM** [comment]
# Returning `&Vec<u8>` as `js_sys::Uint8Array` from Rust

I wanted to return `js_sys::Uint8Array` from my function. I got it working this way for anyone who might be curious:

From Rust:
```rs
use core::{decoder, byte_reader};
use js_sys::{Array, Uint8Array};

use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub fn decode_raw_bytes() -> Uint8Array {
    let decoded_bytes: &Vec<u8> = decoder.run();
    let a: Array = decoded_bytes.into_iter().map(|num| JsValue::from(*num)).collect();
    return Uint8Array::new(&a);
}
```

Generated TS definition:
```ts
export function decode_raw_bytes(): Uint8Array;
```

Usage from Javascript:
```js
import gluedWasm from './pkg/wasm.js';

const decoded_bytes_in_array = gluedWasm.decode_raw_bytes();
console.log(decoded_bytes_in_array) // Uint8Array(n) [ x, y, z, ... ]
```

**Comment by zimond** [comment]
@9oelM this is `From<&u8>` [implemented](https://docs.rs/js-sys/latest/js_sys/struct.Uint8Array.html#impl-From%3C%26%27a%20%5Bu8%5D%3E) for Uint8Array, so basically you could just return `decoded_bytes.into()`

**Comment by 9oelM** [comment]
@zimond hey, thanks for the cool solution. I appreciate it. I'm VERY new to Rust, so could you tell me a lil more?:

If I change the code that way, the compiler complains like so:
```rs
the trait bound `Uint8Array: From<&Vec<u8>>` is not satisfied
the following implementations were found:
  <Uint8Array as From<&'a [u8]>>
  <Uint8Array as From<JsValue>>
required because of the requirements on the impl of `Into<Uint8Array>` for `&Vec<u8>`
```

<img width="696" alt="image" src="https://user-images.githubusercontent.com/22465806/164410193-16e44cf2-69ba-4376-9dc8-68aa07a6ea89.png">


My suspicion (may be totally wrong) is that it's `&'a [u8]` which is not `&[u8]`. Is that right??

**Comment by zimond** [comment]
@9oelM That's because you explicitly declared that `decoded_bytes` is `&Vec<u8>`. There's another API `Vec::as_slice` which could give you a `&[u8]` from `Vec<u8>`. So you could use `decoded_bytes.as_slice().into()`, or `Uint8Array::from(decoded_bytes.as_slice())`.

More learning materials: [Deref coercions](https://doc.rust-lang.org/book/ch15-02-deref.html#implicit-deref-coercions-with-functions-and-methods), [Type conversion](https://doc.rust-lang.org/rust-by-example/conversion/from_into.html) , [Lifetime ellision](https://doc.rust-lang.org/reference/lifetime-elision.html) (which is why `&'a [u8]` and `&[u8]` are the some)

**Comment by 9oelM** [comment]
@zimond wow thanks a bunch!! I will change the code like so then. Sorry about polluting the thread with the qs guys. But this will be it. Thanks again!

**Comment by ZhaoXiangXML** [comment]
> 



> Unfortunately, solutions with `js_sys` have huge performance overheads. You can also return `Box<[T]>` with `T` being _basic numeric type_ which is efficient on the JS side (it just slices relevant wasm memory). This however needs a copy on the Rust side. In general, it is faster than `js_sys` though.
> 
> What I come up with (and seems to be faster by a margin) is to have the following wrapper type:
> 
> ```
> pub struct MySlice<T> {
>     phantom: std::marker::PhantomData::<T>,
>     _ptr: u32,
>     _len: u32,
> }
> 
> impl<T: wasm_bindgen::describe::WasmDescribe> wasm_bindgen::describe::WasmDescribe for MySlice<T> {
>     fn describe() {
>         wasm_bindgen::describe::inform(wasm_bindgen::describe::REF);
>         wasm_bindgen::describe::inform(wasm_bindgen::describe::SLICE);
>         T::describe();
>     }
> }
> 
> impl<T: wasm_bindgen::describe::WasmDescribe> wasm_bindgen::convert::IntoWasmAbi for MySlice<T> {
>     type Abi=wasm_bindgen::convert::WasmSlice;
> 
>     #[inline]
>     fn into_abi(self) -> wasm_bindgen::convert::WasmSlice {
>         wasm_bindgen::convert::WasmSlice {
>             ptr: self._ptr,
>             len: self._len,
>         }
>     }
> }
> 
> impl<T> std::convert::From<&Vec<T>> for MySlice<T> {
>     fn from(vec: &Vec<T>) -> Self {
>         let _ptr = vec.as_ptr() as u32;
>         let _len = vec.len() as u32;
>         
>         Self {
>             phantom: std::marker::PhantomData,
>             _ptr,
>             _len
>         }
>     }
> }
> ```
> 
> This is both zero-copy on the rust side and single copy between wasm memory and JS on the JS side. Basically, the type just emulates ref to a slice through standard wasm bindgen API (which means support for returning `&[T]` should be relatively easy to add if somebody knows internals of bindgen) A caveat is that you _shouldn't_ save MySlice, it should only be used to return immediate data to JS (because it detaches pointers, the code cannot guarantee lifetimes)
> 
> Usage is simple:
> 
> ```
> #[wasm_bindgen]
> pub struct Container {
>     data: Vec<f64>,
> }
> 
> #[wasm_bindgen]
> impl Container {
>     pub fn read(&self) -> MySlice<f64> {
>         (&self.data).into()
>     }
> }
> ```
> 
> with the JS side using
> 
> ```
> const c = Container.new() // somehow create
> const values = c.read()
> ```

If `T` is a rust struct, I got this in the generated typescript:

```typescript
/**
*/
export class Container {
  free(): void;
/**
* @returns {any[]}
*/
  read(): any[];
}
```

Is it possible to have T[] as return type?

**Comment by zimond** [comment]
@ZhaoXiangXML https://rustwasm.github.io/wasm-bindgen/reference/attributes/on-rust-exports/typescript_type.html

**Comment by ZhaoXiangXML** [comment]
> @ZhaoXiangXML https://rustwasm.github.io/wasm-bindgen/reference/attributes/on-rust-exports/typescript_type.html

I end up using a macro to impl the slice types, with `#[wasm_bindgen]` and member functions `length()`, `get()` and `set()`.

**Comment by bushrat011899** [comment]
Is there any work on this issue still? I saw #2734 had a solution to this issue but needed additional help with passing tests. If a full resolution cannot be implemented without a breaking change, could the `#[wasm_bindgen]` macro be extended with an attribute similar to `getter_with_clone` which uses serialization to coerce function outputs into something JS compatible?

```rust
/// Current Workaround
#[wasm_bindgen]
pub fn manual_serialization() -> JsValue {
    serde_wasm_bindgen::to_value::<Vec<String>>(&vec![]).unwrap()
}

/// Proposed Alternative
#[wasm_bindgen(serialize_output)]
pub fn automatic_serialization() -> Vec<String> {
    vec![]
}
```

This appears to be the workaround I've seen the most commonly used, but has a massive flaw in that all type information is lost for the developer and `wasm-bindgen` itself for type annotations. I saw that #2649 was closed because of the potential hidden performance impact from a solution like this. I would think that adding this as a macro option would allow the ergonomics without hiding the issues.

**Comment by sinking-point** [comment]
@bushrat011899

I have had no response from @chinedufn @alexcrichton regarding the failing test. I spent a lot of time on this PR and this is discouraging to say the least. I wonder if they just never look at PRs unless they pass all tests. In which case, I could assume an interpretation of the comment in the test, and get it to pass. That way I'd get their attention, and if I assumed the wrong interpretation they will tell me to change it. However the PR is quite old now and probably has merge conflicts with master to sort out. If I'm going to do more work on this I want some assurance that I'm not wasting my time.

**Comment by bushrat011899** [comment]
> @bushrat011899
> 
> I have had no response from @chinedufn @alexcrichton regarding the failing test. I spent a lot of time on this PR and this is discouraging to say the least. I wonder if they just never look at PRs unless they pass all tests. In which case, I could assume an interpretation of the comment in the test, and get it to pass. That way I'd get their attention, and if I assumed the wrong interpretation they will tell me to change it. However the PR is quite old now and probably has merge conflicts with master to sort out. If I'm going to do more work on this I want some assurance that I'm not wasting my time.

That really is unfortunate, this seems like a really obvious feature to work on in my opinion. My first thought for something useful with WASM was to write a REST API client. That way, I could have type guarantees and unit tests that, for example, my Axum server was interacting with a web UI correctly. Kinda hard to do that if something simple like "give me a list of data" requires type erasure.

**Comment by bushrat011899** [comment]
Because this issue still doesn't have a good workaround IMO, I've tried to make my own. Below is a single-file procedural attribute macro which modifies a function definition in-place to:

1. Return a `JsValue` instead of whatever type it is supposed to return.
2. Evaluate the function body (async or sync) and store in a temporary variable `__result`.
3. Pass the `__result` to `serde_wasm_bindgen` to create a `JsValue` to return.

This doesn't fix the issue entirely. For example, `wasm-bindgen` will mark the TypeScript definition of this function as returning `any` (understandable, it is!), and serialization is not always required to cross the FFI boundary.

I haven't published a crate before, and this is small enough that I don't know if it really needs to be, but this is how I've worked around this pretty massive functionality gap in `wasm-bindgen`. Hope someone finds it useful!

### Proc-Macro Crate `lib.rs`
```rust
use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::{parse::Nothing, parse2, parse_quote, ItemFn, ReturnType, Type};

#[proc_macro_attribute]
pub fn wrap_jsvalue(args: TokenStream, input: TokenStream) -> TokenStream {
    let input: TokenStream2 = input.into();

    let function = parse2::<Nothing>(args.into()).and(parse2::<ItemFn>(input.clone()));

    let expanded = match function {
        Ok(function) => expand_macro(function),
        Err(parse_error) => {
            let compile_error = parse_error.to_compile_error();
            quote!(#compile_error #input)
        }
    };

    expanded.into()
}

fn expand_macro(mut function: ItemFn) -> TokenStream2 {
    let ReturnType::Type(arrow, _return_type) = function.sig.output else {
        return quote!(#function);
    };

    let stmts = function.block.stmts;

    let result = if function.sig.asyncness.is_some() {
        quote!((move || async move { #(#stmts)* })().await)
    } else {
        quote!((move || { #(#stmts)* })())
    };

    function.block = Box::new(parse_quote!({
        let __result = #result;

        serde_wasm_bindgen::to_value(&__result).unwrap()
    }));

    function.sig.output = ReturnType::Type(arrow, Box::new(Type::Verbatim(quote!(JsValue))));

    quote!(#function)
}
```

### Primary Crate `lib.rs`
```rust
use wasm_bindgen::prelude::*;
use my_macros_crate::wrap_jsvalue;

#[wrap_jsvalue]
#[wasm_bindgen]
pub async fn test() -> Vec<String> {
    vec!["Hello", "World"]
}
```

**Comment by fjarri** [comment]
@bushrat011899: I hit this issue (and similar ones with argument types) before, and tried to generalize it as https://docs.rs/wasm-bindgen-derive/latest/wasm_bindgen_derive/ . Would be interesting to include your macro there as well.

**Comment by bushrat011899** [comment]
@fjarri I'm not surprised, it seems like a pretty easy issue to hit! Feel free to take the code I wrote and include it if you like. It's straightforward enough that I'm sure any reasonable person would end up converging on this independently anyway.

**Comment by rossng** [comment]
Great work @Liamolucko! I look forward to using this once released.

**Comment by gabrielDevlog** [comment]
Hi guys, great to see a PR has been merged.
Is there any ETA for this feature ? Or is it possible for you to trigger a new release, so we can have this feature using cargo add.

Thanks

**Comment by daxpedda** [comment]
See https://github.com/rustwasm/wasm-bindgen/issues/3530.

**Comment by samuelOsborne** [comment]
@daxpedda I checked that issue, still not released right ?

**Comment by daxpedda** [comment]
No.
You can find all our releases in the [release section](https://github.com/rustwasm/wasm-bindgen/releases).

**Comment by Teebor-Choka** [comment]
When's the next release, the last one was in June.

**Comment by daxpedda** [comment]
We don't have a timetable for a next release.
See #3530.

**Comment by mnjdhl** [comment]
I am also stuck with this issue:
![image](https://github.com/rustwasm/wasm-bindgen/assets/39954681/6ad16c3a-a1bc-4c1f-ba87-660c3b6a87f3)

Has the fix been released already?

**Comment by mnjdhl** [comment]
Even it doesn't work with vector of basic types such as Vec<i32>:
![image](https://github.com/rustwasm/wasm-bindgen/assets/39954681/9bc12990-08a1-480c-bb83-aed0de25b50b)

**Comment by daxpedda** [comment]
> Has the fix been released already?

Yes it has been.
